R/qgis-configure.R

```diff
@@ -87,131 +108,184 @@ qgis_configure <- function(quiet = FALSE, use_cached_data = FALSE) {
       glue("cache-{version}.rds")
     )
 
+    # Practically all code of this function now consists of handling
+    # use_cached_data = TRUE. This includes cases where qgis_reconfigure() must
+    # be called, when a condition about the cache is not met. Note that
+    # qgis_configure() by default (use_cached_data = FALSE) calls
+    # qgis_reconfigure() straight away.
+
     if (use_cached_data && file.exists(cache_data_file)) {
       try({
         cached_data <- readRDS(cache_data_file)
 
         if (!quiet) message(glue("Checking configuration from '{cache_data_file}'"))
 
-        # respect environment variable/option for path
+        # CACHE CONDITION: contains minimum required elements (i.e. objects; the
+        # cache is an environment)
+        if (
+          !all(
+            c("path", "version", "algorithms", "plugins", "use_json_output")
+            %in% names(cached_data)
+          )
+        ) {
+          message(
+            "The cache does not contain all required elements.\n",
+            "Will try to reconfigure qgisprocess and build new cache ..."
+          )
+          qgis_reconfigure(cache_data_file = cache_data_file, quiet = quiet)
+          return(invisible(has_qgis()))
+        }
+
+        # CACHE CONDITION: the path element does not contradict the environment
+        # variable/option for the qgis_process path
         option_path <- getOption(
           "qgisprocess.path",
           Sys.getenv("R_QGISPROCESS_PATH")
         )
 
-        if (identical(option_path, "") || identical(option_path, cached_data$path)) {
-
-          if (!quiet) message(glue(
-            "Checking cached QGIS version with version reported by '{cached_data$path}' ..."
+        if (!identical(option_path, "") && !identical(option_path, cached_data$path)) {
+          message(glue(
+            "The user's qgisprocess.path option or the R_QGISPROCESS_PATH environment ",
+            "variable specify a different qgis_process path ({option_path}) ",
+            "than the cache did ({cached_data$path}).\n",
+            "Hence rebuilding cache to reflect this change ..."
           ))
+          qgis_reconfigure(cache_data_file = cache_data_file, quiet = quiet)
+          return(invisible(has_qgis()))
+        }
+
+        # CACHE CONDITION: the elements checked by has_qgis() look OK
+
+        condition <-
+          is.character(cached_data$version) &&
+          length(cached_data$version) == 1L &&
+          nchar(cached_data$version) > 0L &&
+          is.character(cached_data$path) &&
+          length(cached_data$path) == 1L &&
+          nchar(cached_data$path) > 0L &&
+          inherits(cached_data$algorithms, "data.frame") &&
+          inherits(cached_data$plugins, "data.frame")
```

Comment:
This condition only cares about the form of the cache objects, before loading the cache, so there are no new values involved. Checking and loading an existant cache is now almost the entire focus of the code of `qgis_configure()`; rebuilding the cache was transferred to the internal `qgis_reconfigure()`.

However it can be much simplified, since **assertthat** was imported recently; notably `is.string()` can be used to replace the first six subconditions here.

Reply:
067aec3